lib/spack/spack/relocate.py

```diff
@@ -413,92 +495,75 @@ def replace(match):
+    env_rpaths = spack.build_environment.get_rpaths(spec.package)
+
     for path_name in path_names:
         if path_name.endswith('.o'):
             continue
         if new_dir == old_dir:
             continue
         if platform.system().lower() == 'darwin':
             rpaths, deps, idpath = macho_get_paths(path_name)
-            # one pass to replace placeholder
-            (n_rpaths,
-             n_deps,
-             n_idpath) = macho_replace_paths(placeholder,
-                                             new_dir,
-                                             rpaths,
-                                             deps,
-                                             idpath)
-            # another pass to replace old_dir
             (new_rpaths,
              new_deps,
-             new_idpath) = macho_replace_paths(old_dir,
-                                               new_dir,
-                                               n_rpaths,
-                                               n_deps,
-                                               n_idpath)
+             new_idpath) = macho_replace_paths(old_dir, new_dir,
+                                               rpaths, deps, idpath,
+                                               env_rpaths)
             modify_macho_object(path_name,
                                 rpaths, deps, idpath,
                                 new_rpaths, new_deps, new_idpath)
         else:
-            modify_object_macholib(path_name, placeholder, new_dir)
-            modify_object_macholib(path_name, old_dir, new_dir)
-        if len(new_dir) <= len(old_dir):
-            replace_prefix_bin(path_name, old_dir,
-                               new_dir)
-        else:
-            tty.warn('Cannot do a binary string replacement'
-                     ' with padding for %s'
-                     ' because %s is longer than %s' %
-                     (path_name, new_dir, old_dir))
+            rpaths, deps, idpath = macholib_get_paths(path_name)
+            (new_rpaths,
+             new_deps,
+             new_idpath) = macho_replace_paths(old_dir, new_dir,
+                                               rpaths, deps, idpath,
+                                               env_rpaths)
+            modify_object_macholib(path_name, rpaths=rpaths, deps=deps,
+                                   idpath=idpath, old_prefix=old_dir,
+                                   new_rpaths=new_rpaths, new_deps=new_deps,
+                                   new_idpath=new_idpath, new_prefix=new_dir)
 
 
-def relocate_elf_binaries(path_names, old_dir, new_dir, allow_root):
+def relocate_elf_binaries(path_names, orig_rpaths, spec):
     """
     Change old_dir to new_dir in RPATHs of elf binaries
     Account for the case where old_dir is now a placeholder
     """
-    placeholder = set_placeholder(old_dir)
+    env_rpaths = spack.build_environment.get_rpaths(spec.package)
+    new_rpaths = env_rpaths
```

Comment:
`env_rpaths` doesn't appear to be used other than to assign to `new_rpaths` - can you collapse these two lines into one?

Also, if this is how all of the new rpaths are determined, it might miss some rpaths: `build_environment.get_rpaths` doesn't make use of the custom `libs` implementations of some packages, which may define additional library dirs.

Reply:
Would those additional rpaths be added to the rpath in the existing binaries? This would be a case where the length of the old rpaths would not match the length of the rpaths taken from the build_environment and finding paths based on hashes would need to be done.